external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSerdeSuite.scala

```diff
@@ -144,6 +173,43 @@ object AvroSerdeSuite {
   private val CATALYST_STRUCT =
     new StructType().add("foo", new StructType().add("bar", IntegerType))
 
+  /**
+   * Specifier for type of serde to be used for easy creation of tests that do both
+   * serialization and deserialization.
+   */
+  object SerdeType extends Enumeration {
+    type SerdeType = Value
+    val SERIALIZE, DESERIALIZE = Value
+  }
+  import SerdeType._
+
+  /**
+   * Specifier for type of field matching to be used for easy creation of tests that do both
+   * positional and by-name field matching.
+   */
+  object FieldMatchType extends Enumeration {
+    type FieldMatchType = Value
+    val BY_NAME, BY_POSITION = Value
+  }
+  import FieldMatchType._
+
+  private def createSerde(
+      catalystSchema: StructType,
+      avroSchema: Schema,
+      serdeType: SerdeType,
+      fieldMatchType: FieldMatchType): Either[AvroSerializer, AvroDeserializer] = {
+    val positional = fieldMatchType match {
+      case BY_NAME => false
+      case BY_POSITION => true
+    }
+    serdeType match {
+      case SERIALIZE => Left(new AvroSerializer(catalystSchema, avroSchema, false, positional,
```

Comment:
QQ: why  using Left/Right here?

Reply:
This is a way for `createSerde` to return either `AvroSerializer` or `AvroDeserializer`, depending on the input value (`SERIALIZE` vs `DESERIALIZE`).

I realized a cleaner way to do this is to pass the serializer/deserializer as a factory object which is properly typed. Then the caller doesn't have to extract the proper side of the `Either`. I updated the PR.